mersenne-31/src/x86_64_avx2/poseidon2.rs

```diff
@@ -1,58 +1,290 @@
-use p3_poseidon2::{matmul_internal, DiffusionPermutation};
-use p3_symmetric::Permutation;
+use alloc::vec::Vec;
+use core::arch::x86_64::{self, __m256i};
 
-use crate::{
-    DiffusionMatrixMersenne31, Mersenne31, PackedMersenne31AVX2, POSEIDON2_INTERNAL_MATRIX_DIAG_16,
-    POSEIDON2_INTERNAL_MATRIX_DIAG_24,
+use p3_poseidon2::{
+    external_initial_permute_state, external_terminal_permute_state, sum_15, sum_23, ExternalLayer,
+    ExternalLayerConstants, ExternalLayerConstructor, InternalLayer, InternalLayerConstructor,
+    MDSMat4,
 };
 
-impl Permutation<[PackedMersenne31AVX2; 16]> for DiffusionMatrixMersenne31 {
-    fn permute_mut(&self, state: &mut [PackedMersenne31AVX2; 16]) {
-        matmul_internal::<Mersenne31, PackedMersenne31AVX2, 16>(
-            state,
-            POSEIDON2_INTERNAL_MATRIX_DIAG_16,
-        );
+use crate::{exp5, Mersenne31, PackedMersenne31AVX2, P, P_AVX2};
+
+/// The internal layers of the Poseidon2 permutation for Mersenne31.
+///
+/// This is optimized for the AVX2 architecture.
+#[derive(Debug, Clone)]
+pub struct Poseidon2InternalLayerMersenne31 {
+    pub(crate) internal_constants: Vec<Mersenne31>,
+    packed_internal_constants: Vec<__m256i>,
```

Comment:
Why `Vec`s instead of arrays?

Reply:
The issue with array's would be that it's not completely clear how to determine the length at compile time in a nice way. The only way I could see to do this would involve adding two more constant parameters which would percolate through the code. 

These are also generated when the instance is initialized so there shouldn't be a runtime cost.